HW2/P2/parallel_vector.pyx

```diff
@@ -79,21 +82,58 @@ cpdef move_data_fine_grained(np.int32_t[:] counts,
    # Use parallel.prange() and a lock for each element of counts to parallelize
    # data movement.  Be sure to avoid deadlock, and double-locking.
    ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
+
+   for r in prange(repeat, nogil=True, num_threads=threads): #comment to make work
+      
+      
+   #for r in range(repeat): #uncomment to make work
+      for idx in range(src.shape[0]):
+        
+        
+         if counts[src[idx]] > 0:
+
+             if src[idx] > dest[idx]: #this line prevents deadlock
+                 #ordering lock aquiring and release prevents the deadlock
+                 acquire(&locks[src[idx]])
+                 acquire(&locks[dest[idx]])
+
+                 counts[dest[idx]] += 1
+                 counts[src[idx]] -= 1
+
+                 release(&locks[src[idx]])
+                 release(&locks[dest[idx]])
+
+             elif src[idx] < dest[idx]: #this line prevents deadlock
+
+                 acquire(&locks[dest[idx]])
+                 acquire(&locks[src[idx]])
+
+                 counts[dest[idx]] += 1
+                 counts[src[idx]] -= 1
+
+                 release(&locks[src[idx]])
+                 release(&locks[dest[idx]])
+
+             else: #this condition prevents double locking
+
+                 acquire(&locks[dest[idx]])
+
+                 counts[dest[idx]] += 1
+                 counts[src[idx]] -= 1
+
+                 release(&locks[dest[idx]])
+
```

Comment:
Good thinking about the locking order. It may be advantageous to split source and destination locking, since you only work on one resource at a time and other threads cannot access the other resource in the mean time.